### PR TITLE
127 customize gateway from bosconfigjson

### LINKED ISSUE #127: Customize gateway from bos.config.json
bos-workspace supports a custom gateway via a -g flag and either a local file path, or a url to nearfs bundle or CDN. We want to be able to configure this from bos.config.json

Add a "gateway" section to the bos.config.json

I'm thinking something like this:

```json
{
  "gateway": {
    "tagName": "near-social-viewer", // element tag
    "bundleUrl": "https://ipfs.web4.near.page/ipfs/bafybeifvti4qqgrjnyzdzueitruup6rnqwasxt742stdyby2xey3ymnhie", // path to dist
  }
}
```

Although, also considering a version like this:

```
{
  "gateway": {
    "tag": "near-social-viewer", // element tag
    "source": "https://ipfs.web4.near.page/ipfs/bafybeifvti4qqgrjnyzdzueitruup6rnqwasxt742stdyby2xey3ymnhie", // path to dist
    "version": "1.0.0", // needed?
    "dependencies": {}, // script tags to include
    "type": "" // type definition
  }
}
```

## Acceptance Criteria
- [x] gateway is optional in config
- [x] -g flag should take precedence, even if config is provided
- [x] should use the bundleUrl in modifyIndexHtml if provided
- [x] should use the tagName in modifiyIndexHtml if provided
- [x] bundleUrl supports a trailing slash or no trailing slash
- [x] bundleUrl supports a relative path to local dist (../../near-bos-webcomponent/dist)
- [x] throws error if tagName but no bundleUrl, or bundleUrl but no tagName

### PR DESCRIPTION
**Pull request by bb-face**
Resolves #127

### CODE REVIEW

File: lib/dev.ts
Comment (elliotBraem) [inline lib/dev.ts]: We can remove this from devMulti -- it gets complicated fast. Let's only do a single dev app.

devMulti currently used DEFAULT_LOCAL_GATEWAY_URL and we should keep it that way.

File: lib/server.ts
Comment (elliotBraem) [inline lib/server.ts]: I think there is a cleaner solution with just DEFAULT_GATEWAY object with these as default values, instead of having to maintain all these exports + constants

File: lib/dev.ts
Comment (elliotBraem) [inline lib/dev.ts]: I'm thinking about this:

Since bundleUrl + tagName are an essential combo, and -g ${url} in commands has been a bit complicated, we can treat the gateway configuration in bos.config.json as the only place to configure and as source of truth. 

We do need to handle --no-gateway, which Commander.js (cli-tool) automatically assigns to the "gateway" variable in dev options. 

In short, I'm thinking gateway should be only boolean that handles whether it is enabled or not.

Then bos.config.json gateway handles the bundleUrl and tagName (and does not have enabled field, otherwise will have conflicts and unexpected behavior)

Comment (elliotBraem) [reply lib/dev.ts]: So we will change README to reflect and say customizing gateway happens in bos.config.json

I'm not worried about backwards compatibility here, because only our projects are using custom gateway atm.

Comment (bb-face) [reply lib/dev.ts]: So the requirements are:

1) If -g is false or there's the option --no-gateway, gateway is disabled;
2) if -g is true and there's a gateway configuration in `bos.config.json` takes the values from there;
3) if -g is true but there's no gateway configuration in `bos.config.json` use default values;

Question: shall we still make both `bundleUrl` and `tagName` both mandatory in `bos.config.json`?

File: README.md
Comment (elliotBraem) [inline README.md]: "To use this feature, specify the gateway bundle url and the tag name in the `bos.config.json` file."

File: README.md
Comment (elliotBraem) [inline README.md]: And then this can be the sample json addition

```json
	"gateway": {
		"bundleUrl": "https://ipfs.web4.near.page/ipfs/bafybeibe63hqugbqr4writdxgezgl5swgujay6t5uptw2px7q63r7crk2q/",
		"tagName": "near-social-viewer"
	}
```

File: lib/config.ts
Comment (elliotBraem) [inline lib/config.ts]: Do we need to call it an Object? Why not just GatewayConfig?

File: lib/dev.ts
Comment (elliotBraem) [inline lib/dev.ts]: -g would be true by default, we can remove "gateway" from our cli actually

Only thing would be if user provides --no-gateway, which will disable it

File: lib/dev.ts
Comment (elliotBraem) [inline lib/dev.ts]: Considering we have a default config, ability to override this config, or --no-gateway to disable it, I'm not entirely sure what this function does anymore. Take a step back and reassess it's purpose and if it's clearer to just check if gateway is disabled or not

Although we do need error checking -- we can't have bundleUrl without tagName and can't have tagName without bundleUrl

Comment (elliotBraem) [reply lib/dev.ts]: (and let's have a test to confirm both must be populated)

File: lib/server.ts
Comment (elliotBraem) [inline lib/server.ts]: We don't need the = DEFAULT _GATEWAY since we know startDevServer already has this fallback

File: lib/server.ts
Comment (elliotBraem) [inline lib/server.ts]: We can remove these notes

File: lib/server.ts
Comment (elliotBraem) [inline lib/server.ts]: I think this `const gatewayInitPromise =` would introduce strange behavior? Does the web server stutter at all when starting?

I thought gatewayInitPromise is a constant already defined in this file and initializeGateway function was setting it

In other words, I don't think you should have this const gatewayInitPromise here

File: lib/server.ts
Comment (elliotBraem) [inline lib/server.ts]: Does this work? Isn't gateway an object? You mean gateway.bundleUrl?

Confirm by running the command with `-l DEBUG` to turn on logging

File: lib/server.ts
Comment (elliotBraem) [inline lib/server.ts]: Following up on previous comment, we don't need the return. Should be as it was before

gatewayInitPromise = setupGateway

File: lib/server.ts
Comment (elliotBraem) [inline lib/server.ts]: This isn't right, I think you mean gateway.enabled

File: tests/unit/gateway.ts
Comment (elliotBraem) [inline tests/unit/gateway.ts:77]: Can we append to these tests to make sure this uses .tagName?

File: tests/unit/server.ts
Comment (elliotBraem) [inline tests/unit/server.ts]: @/lib/dev

**Comment by elliotBraem** [comment]
@race-of-sloths include

**Comment by race-of-sloths** [comment]
@bb-face Thank you for your contribution! Your pull request is now a part of the [Race of Sloths](https://race-of-sloths.com)!


[<picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://badge.race-of-sloths.com/bb-face?type=bot&theme=dark">
    <source media="(prefers-color-scheme: light)" srcset="https://badge.race-of-sloths.com/bb-face?type=bot&theme=light">
    Shows profile picture for the author of the PR
</picture>](https://race-of-sloths.com/profile/bb-face)





<details><summary>Current status: <i>executed</i></summary>

| Reviewer | Score |
|--------|--------|
| @elliotBraem  | 5 |


Your contribution is much appreciated with a final score of 5!
You have received 70 (50 base + 20 weekly bonus) Sloth points for this contribution


Another weekly streak completed, well done @bb-face! To keep your weekly streak and get another bonus make pull request next week! Looking forward to see you in race-of-sloths
</details><details><summary>What is the Race of Sloths</summary>
<p>
Race of Sloths is a friendly competition where you can participate in challenges and compete with other open-source contributors within your normal workflow

For contributors:

- Tag @race-of-sloths inside your pull requests
- Wait for the maintainer to review and score your pull request
- Check out your position in the [Leaderboard](https://race-of-sloths.com/leaderboard)
- Keep weekly and monthly streaks to reach higher positions
- Boast your contributions with a dynamic picture of your [Profile](https://race-of-sloths.com/profile/bb-face)

For maintainers:
- Score pull requests that participate in the Race of Sloths
- Engage contributors with fair scoring and fast responses so they keep their streaks
- Promote the Race to the point where the Race starts promoting you
- Grow the community of your contributors

Feel free to check [our website](https://race-of-sloths.com) for additional details!
</p>
</details> 

<details>
<summary>Bot commands</summary>

- For contributors
  - **Include a PR:** `@race-of-sloths include` to enter the Race with your PR
- For maintainers:
  - **Assign points:** `@race-of-sloths score [1/2/3/5/8/13]` to award points based on your assessment.
  - **Reject this PR:** `@race-of-sloths exclude` to send this PR back to the drawing board.
  - **Exclude repo:** `@race-of-sloths pause` to stop bot activity in this repo until `@race-of-sloths unpause` command is called

</details>

File: lib/dev.ts
Comment (bb-face) [reply lib/dev.ts]: The error checking is in `config.ts` in the configuration schema, if I try to run the app without a gateway with both `tagName` and `bundleUrl` it will return this error: 

```
(node:20505) [DEP0040] DeprecationWarning: The `punycode` module is deprecated. Please use a userland alternative instead.
(Use `node --trace-deprecation ...` to show where the warning was created)
✗ SyntaxError: bos.config.json: Expected double-quoted property name in JSON at position 421 (line 14 column 2)
    at JSON.parse (<anonymous>)
    at _readFile (/Users/stefano/Workspace/Work/NEAR/bos-workspace/node_modules/jsonfile/index.js:25:16)
```

Without the option `-g` the logic is simpler but still needs to be expressed somewhere, that's the purpose of that function, somehow we need to specify the logic of the gateway: if it's disabled, if it's enabled but without a configuration object in `bos.config.json` or if it's enabled with a correctly formatted configuration object!

Let me know if you want to introduce more error checking on top of the Joi schema or if I need to move the gateway logic somewhere else!

**Comment by elliotBraem** [comment]
@race-of-sloths score 5

**Review by elliotBraem — APPROVED** [review]

**Pull request merged**
